경북대 BE_박준희 2단계 - 주문하기

jjunii opened the pull request:
# 2단계 주문하기
## 구현 기능
- 주문하기 기능
    - 주문 생성 및 저장
    - 상품 옵션의 수량 차감
    - 위시 리스트에서 해당 상품 삭제
- 카카오 API 연동 기능
    - 주문 내역 카카오톡 메세지로 전송
   
## 리뷰 반영
- env.yaml 파일을 프로젝트 루트 경로로 이동
- URI 생성을 KakaoService로 위임 & 앱 실행 시 한 번만 생성

## 질의 사항
주문하기 로직에 대해서, 제 생각에는 메세지가 발송되지 않더라도 주문은 정상적으로 처리되어야 한다고 생각했습니다.
하지만 서비스 로직에서 `placeOrder()`에 트랜잭션을 적용하니, 메세지가 발송되지 않으면 주문도 롤백되는 문제가 있었습니다.
이를 해결하기 위해 주문하기과 메세지 보내기를 분리해야겠다고 생각했습니다.

처음에는 OrderService내에서 `saveOrder()`를 트랜잭션이 붙은 독립적인 메서드로 분리하고,
`placeOrder()`에서 `saveOrder()`를 호출하는 방법을 시도했지만,
같은 클래스 내부 호출이라 self-invocation 문제가 발생했습니다다.

그래서 이를 해결하기 위해 두 가지 방법을 고려해보았습니다.
1. 주문하기 로직을 별도 클래스로 분리
: OrderDbService에 DB 접근 로직(`saveOrder()`)을 구현하고,
OrderService에서 OrderDbService와 KakaoMessageService 호출
2. 컨트롤러에서 주문하기와 메세지 보내기 분리
: 컨트롤러에서 OrderService(주문하기)와 KakaoMessageService(메세지 보내기) 순서대로 실행

이 중에서 현재는 2번 방식을 선택하였습니다.
서비스 간 의존보다는 컨트롤러가 필요한 서비스를 호출하는 것이 계층형 아키텍처에 더 적합하다고 판단했기 때문입니다.
다만, 이렇게 구현하다 보니 컨트롤러에서 엔티티를 직접 다루게 되고, 컨트롤러의 역할이 많아진 것 같습니다.
이런 구조가 괜찮은지, 더 나은 설계 방식이 있다면 조언 부탁드립니다!

jjunii committed:
refactor: env 파일을 프로젝트 루트 경로로 이동

jjunii committed:
refactor: 카카오 연동 URL 생성 로직을 Service로 위임

jjunii committed:
feat: 카카오 회원가입을 위한 Member 엔티티 구조 변경

jjunii committed:
feat: 카카오 사용자 정보 조회 (고유 ID -> 이메일) 변경

jjunii committed:
feat: 카카오 로그인/회원가입 및 JWT 인증 구현

- 조회한 카카오 사용자 정보를 기반으로 Member DB에 회원가입 또는 로그인
- 동일 이메일로 카카오 로그인/회원가입 또는 로컬 로그인/회원가입 중복 시 예외 처리
- 인증된 사용자에게 JWT 발급 및 반환

jjunii committed:
docs(readme): 구현 목록 추가

jjunii committed:
feat: Order 관련 엔티티 및 DTO 생성

jjunii committed:
feat: 주문하기 기능 구현

jjunii committed:
feat: 주문 내역 카카오 메세지 전송 구현

sumniy commented on src/main/java/gift/controller/KakaoAuthController.java:
```diff
@@ -1,43 +1,32 @@
-
         HttpHeaders headers = new HttpHeaders();
-        headers.setLocation(uri);
+        headers.setLocation(kakaoAuthService.getKakaoAuthUrl());
```
서비스 계층으로의 책임 분리 👍

sumniy commented on src/main/java/gift/controller/OrderController.java:
```diff
@@ -0,0 +1,52 @@
+
+        Order savedOrder = orderService.placeOrder(orderRequestDto);
+
+        if (savedOrder.getMember().getSignupType() == SignupType.KAKAO) {
+            try {
+                kakaoMessageService.sendOrderMessage(savedOrder);
+            } catch (Exception e) {
+                log.error("카카오 메시지 전송 실패. 주문 ID: {}", savedOrder.getId(), e);
+            }
+        }
```
컨트롤러가 많은 역할을 하고 있는 것 같네요. 롤백되는 것이 문제라면 placeOrder 에서 sendOrderMessage 를 try catch 로 감싸면 해결되지 않나요? OrderService에서 Entity가 반환되는 것도 비즈니스 로직이 컨트롤러 계층으로 올라왔기 때문이라고 생각합니다.

보통 메시지 발송과 같은 별개 비즈니스 로직이 강결합 되는 것을 피하기 위해서 이벤트 방식을 활용하기도 하는데요. 
관련해서는 아래 글을 참고해 보시면 좋을 것 같습니다! 
https://tecoble.techcourse.co.kr/post/2020-09-30-event-publish/

다만 현재 규모에서 이벤트 방식을 적용하는 것은 오히려 오버엔지니어링이라고 생각되니 학습을 목적으로 원하시면 시도해 보시고 아니면 단순하게 하셔도 괜찮습니다~

sumniy commented on src/main/java/gift/dto/OrderRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+import jakarta.validation.constraints.Size;
+
+public record OrderRequestDto(
+        @NotNull(message = "옵션 ID는 필수입니다.")
```
유효성 검사 👍

sumniy commented on src/main/java/gift/dto/OrderRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+        @Min(value = 1, message = "수량은 1개 이상이어야 합니다.")
+        Integer quantity,
+
+        @Size(max = 500, message = "메세지는 최대 500자까지 입력할 수 있습니다.")
```
`max = 500` 
이런 값은 order entity의 제약 조건과 같을 텐데요. 
Order 클래스에서 public static 상수로 선언하고 재사용하는 방법도 있습니다!

sumniy commented on src/main/java/gift/entity/Member.java:
```diff
@@ -18,25 +18,45 @@ public class Member {
+        this.signupType = signupType;
+    }
+
+    public static Member createLocalMember(String email, String password, MemberRole role) {
+        return new Member(null, email, password, role, SignupType.LOCAL);
+    }
+
+    public static Member createKakaoMember(String email, MemberRole role) {
+        return new Member(null, email, null, role, SignupType.KAKAO);
+    }
```
static factory method 활용 좋네요 👍

sumniy commented on src/main/java/gift/service/KakaoAuthService.java:
```diff
@@ -2,26 +2,44 @@
 
-    public String getAccessToken(String code) {
+    public KakaoTokenResponseDto getToken(String code) {
         String url = "https://kauth.kakao.com/oauth/token";
```
`https://kauth.kakao.com`도 base Url 정보로 관리하면 좋을 것 같네요~

jjunii replied:
먼저, 피드백 감사합니다!

다만, '**Base Url 정보로 관리하면 좋겠다**'고 해주신 부분에 대해 확인하고 싶은 점이 있습니다.
이전 단계 리뷰에서 '**클라이언트를 선언할 때 공통되는 부분들은 미리 설정하는 방식**'에 대해 알려주셔서 이 방식에 대해 말씀하시는 건가 싶었는데요,
제가 확인해보니 KakaoAuthService 내부에서는 카카오 API가
- 토큰 발급용: `https://kauth.kakao.com`
- 사용자 정보 조회용: `https://kapi.kakao.com`
이렇게 서로 다른 Base URL을 사용하고 있었습니다.

혹시 이 두 URL을 코드에 직접 작성하는 대신, `application.yaml`에 각각 정의하고 `@Value` 주입받아 사용하는 방식을 말씀하시는 걸까요?

제 이해가 맞는지 확인해주시면 감사하겠습니다!

sumniy replied:
여러 api를 연동하면서 변하지 않는 base url 부분만 (https://kauth.kakao.com) 상수로 관리하고 재사용하는 형태를 말씀 드렸습니다! 이 경우에는 2가지 base url을 관리하는 게 되겠네요!

sumniy commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,95 @@
+        String url = "https://kapi.kakao.com/v2/api/talk/memo/default/send";
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(HttpHeaders.AUTHORIZATION, member.getKakaoAccessToken());
```
accessToken이 만료되었는지에 대한 검증과 만료되었다면 refresh 토큰으로 재발급하는 프로세스가 있어야할 것 같은데요. 
어떻게 구현하면 좋을지 고민해 보시고 수정 부탁드리겠습니다!

sumniy commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,95 @@
+                  .toBodilessEntity();
+    }
+
+    private String createMessage(Order order) {
```
이 메서드는 KakaoMessageDto 에서 static factory method로 만들어도 좋아보입니다!

sumniy commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -0,0 +1,95 @@
+        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
+        body.add("template_object", createMessage(order));
+
+        restClient.post()
+                  .uri(url)
+                  .header("Authorization",
+                          "Bearer " + member.getKakaoAccessToken())
+                  .body(body)
+                  .retrieve()
+                  .toBodilessEntity();
```
body와 token만 받는 메서드로 따로 분리하면 재사용성이 좋은 구조로 될 수 있습니다!

sumniy commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -47,13 +48,36 @@ public TokenResponseDto loginMember(MemberRequestDto memberRequestDto) {
         Member member = memberRepository.findByEmail(memberRequestDto.email())
                                         .orElseThrow(LoginFailedException::new);
 
+        if (member.getSignupType() == SignupType.KAKAO) {
```
Member 도메인에서 책임지도록 하면 더 깔끔한 코드가 될 것 같아요!
`member.checkLocalMember()` 이런 느낌으로 어떨까요~?

sumniy commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -47,13 +48,36 @@ public TokenResponseDto loginMember(MemberRequestDto memberRequestDto) {
+    public TokenResponseDto kakaoLogin(String kakaoEmail,
+            String accessToken, String refreshToken) {
+        Member member = memberRepository.findByEmail(kakaoEmail)
+                                        .orElseGet(() -> Member.createKakaoMember(kakaoEmail,
```
orElseGet 활용 💯

sumniy reviewed (changes requested):
준희님 2단계 요구사항 잘 구현해 주셨네요!
질문에 대한 답과 수정 요청 드릴 부분들을 코멘트로 남겼으니 확인해 주세요!

jjunii committed:
refactor: 도메인 제약 조건을 상수로 분리하여 관리 개선

- 하드코딩된 제약 조건 값들을 해당 엔티티 내부에 public static final 상수로 정의
- MESSAGE_MAX_LENGTH 같이 의미 있는 이름으로 상수화하여 가독성 및 유지보수성 향상
- 관련 DTO 및 Entity 필드에서 해당 상수를 재사용하도록 변경

jjunii committed:
refactor: OrderService 내에서 메시지 전송하도록 호출 구조 개선

- OrderController에서 주문하기와 메시지 전송을 분리 호출하던 구조 개선
- 컨트롤러의 책임 축소 및 호출 흐름 단순화

jjunii committed:
refactor: signupType 확인 로직을 Member 도메인으로 위임

jjunii committed:
refactor: KakaoMessageService에 있던 메세지 생성 로직을 정적 팩토리 메서드로 분리

jjunii committed:
refactor: 메세지 전송 로직에서 HTTP 요청 로직 분리

jjunii committed:
feat: 메시지 전송 중 액세스 토큰 만료 시 리프레시 토큰으로 재발급 및 재전송 처리

sumniy commented on src/main/java/gift/dto/OptionRequestDto.java:
```diff
@@ -9,7 +10,8 @@
 public record OptionRequestDto(
         @NotBlank(message = "옵션명은 필수입니다.")
-        @Size(max = 50, message = "옵션명은 공백을 포함하여 최대 50자까지 입력할 수 있습니다.")
+        @Size(max = Option.OPTION_NAME_MAX_LENGTH
```
피드백 반영 👍

sumniy commented on src/main/java/gift/entity/Member.java:
```diff
@@ -58,6 +58,10 @@ public void updateKakaoTokens(String kakaoAccessToken, String kakaoRefreshToken)
         this.kakaoRefreshToken = kakaoRefreshToken;
     }
+    
+    public boolean checkLocalMember() {
```
강력한 검증을 하는 메서드라면 여기서 throw 해도 괜찮을 것 같아요! 그러면 사용하는 곳에서 좀 더 간단해집니다!

jjunii replied:
멘토님의 피드백을 반영하고자, `checkLocalMember()`에서 직접 예외를 던지는 방안을 고려해 보았습니다.

그런데 현재 이 메서드는 여러 곳에서 서로 다른 예외를 던져야 하는 상황에서 사용되고 있습니다.
- loginMember: `member.checkLocalMember()`가 false 이면 LoginFailedException
- kakaoLogin: `member.checkLocalMember()`가 true 이면 EmailAlreadyExistsException
- sendOrderMessage: `savedOrder.getMember().checkLocalMember()`가 false이면 카카오 메시지 전송

따라서 예외 처리 로직을 `checkLocalMember()` 내부에 두게 되면, 여러 종류의 예외를 한 메서드에서 관리해야 해서 오히려 유연성과 재사용성이 떨어진다고 판단했습니다.
현재처럼 `checkLocalMember()`는 상태를 확인하는 역할만 담당하고, 예외 처리는 호출하는 쪽에서 상황에 맞게 처리하는 방식이  적절하다고 생각합니다!

sumniy commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -32,10 +32,24 @@ public void sendOrderMessage(Order order) {
+                  .toBodilessEntity();
+    }
 
+    private MultiValueMap<String, String> createMessageBody(KakaoOrderMessageDto orderMessage) {
```
restClient의 request body로 api 명세에 맞는 클래스를 정의해서 request dto 객체로 정의하면 좋습니다~

sumniy commented on src/main/java/gift/service/KakaoMessageService.java:
```diff
@@ -34,7 +41,20 @@ public void sendOrderMessage(Order order) {
         KakaoOrderMessageDto orderMessage = KakaoOrderMessageDto.from(order);
 
-        sendRequestToKakao(member.getKakaoAccessToken(), orderMessage);
+        try {
+            sendRequestToKakao(member.getKakaoAccessToken(), orderMessage);
+        } catch (HttpClientErrorException.Unauthorized e) {
+            KakaoTokenResponseDto newToken = kakaoAuthService.refreshAccessToken(
+                    member.getKakaoRefreshToken());
+
+            member.updateKakaoTokens(
+                    newToken.accessToken(),
+                    newToken.refreshToken()
+            );
+            memberRepository.save(member);
+
+            sendRequestToKakao(newToken.accessToken(), orderMessage);
+        }
```
try catch 구문으로 재처리하는 방식으로 구현해 주셨네요! 
다만 요청을 하지 않아도 토큰이 만료되었는지 확인할 수 있는 방법이 있습니다! 
액세스 토큰을 요청할 때 받은 응답(KakaoTokenResponseDto)에 expiresIn 라는 정보가 있는데요. 
이 정보를 검사하면 값비싼 네트워크 요청이 없어도 토큰이 만료 되었는지 바로 확인할 수 있습니다! 
같이 저장해서 활용해보면 좋을 것 같아요!

jjunii replied:
멘토님의 좋은 의견 감사합니다!
`expires_in`을 활용하는 방식에 대해 생각해볼 수 있는 좋은 계기가 되었습니다. 멘토님께서 추천해주신 방향을 고려해보았지만, 제 생각에 현재는 `try-catch` 방식이 더 적합하다고 생각하여 유지하였습니다.

- 토큰의 실제 유효 여부는 카카오 서버만이 정확하게 판단할 수 있기 때문에 서버 간 시간 차이의 문제가 생길 수 있는 사전 계산보다는 카카오의 실제 401 응답을 기준으로 처리하는 방식이 더 신뢰할 수 있다고 생각했습니다.
- 현재는 모든 401 응답을 토큰 만료로 간주하고 있지만, '앱키 오류', '사용자 연결 해제' 등 실패 케이스에 따라 다른 처리를 고려한다면 응답 본문을 기반으로 판단할 수 있는 `try-catch` 방식이 더 유연한 구조라고 생각했습니다.
- 토큰이 유효한 대부분의 경우에는 사전 검사 없이 단 한 번의 요청으로 끝나기 때문에 전반적인 효율이 높다고 생각했습니다.

이와 같은 이유로 현재 구조를 유지했으며, 혹시 제가 미처 고려하지 못한 부분이 있다면 피드백 주시면 감사하겠습니다.
리뷰 덕분에 여러 관점에서 고민해볼 수 있었습니다!

sumniy reviewed (approved):
준희님 2단계 구현과 피드백 반영해 주시느라 고생하셨습니다! 👍
참고하시면 좋을 내용 몇 개 코멘트로 남겼으니 확인해 주세요~
